File: examples/test/testSplitData/testSplitData.m
clear

% Create an object to store data that will be written to the file
% fname.dat in binary format and static memory allocations to fname.h
dat = splitData;

%%
% Generate a matrix-vector multiplication
A = sprandn(100,44,0.1);
b = randn(100,1);
x = randn(44,1);
dat.add_var('x', x);
dat.add_var('sol', A*x+b);

% We create functions that are derived from coderFunc, and then add them to
% the splitData object
f(1) = coderFunc_times('func_blas',       A, 'bConst', b, 'method', 'blas');
f(2) = coderFunc_times('func_forloop',    A, 'bConst', b, 'method', 'for_loops');
f(3) = coderFunc_times('func_sparse',     A, 'bConst', b, 'method', 'sparse');
f(4) = coderFunc_times('func_exhaustive', A, 'bConst', b, 'method', 'exhaustive');
dat.add_function(f);

% You can also #define anything this way
dat.define('n', size(A,1), 'int');

%% Test solution of linear systems of equations
A = sprandn(100,100,0.1); A = A + A';

% Store the test vector and solution
x = randn(100,1);
dat.add_var('x_lin', x);
dat.add_var('y_sol', A\x);

clear f
f(1) = coderFunc_mldivide('func_solve_inv', A, 'method', 'invert');
f(2) = coderFunc_mldivide('func_solve_ldl', A, 'method', 'ldl');

dat.add_function(f);

%% Here's a bunch of examples on how to add variables to a data file
% add_var a bunch of vectors & matrices, giving them names
dat.add_var('aaa',randn(10,1),  'type', 'real', 'desc', 'This is a vector aaa');
dat.add_var('bbb',randn(20,1),'type', 'real');

% Write matrix variables
A = randn(10,3);
dat.add_var('Arow', A','type', 'real'); % Row major format
dat.add_var('Acol', A, 'type', 'real'); % Column major format

% Write sparse matrix
A = sprandn(50,50,0.2);
dat.add_var('Asparse', A,'type', 'real', 'desc', 'This is a sparse matrix');  % Column major format

% Write integer vectors
I = ceil(100*rand(30,1));
dat.add_var('I', I, 'type', 'int');
% Force sparse for a dense matrix
dat.add_var('Isparse', I, 'type', 'int', 'storage_method', 'sparse', 'desc', 'This is a dense integer matrix stored in sparse format');


%% Write the data to file
dat.write_to_file('test');

%% Compile
system('gcc -o test -framework Accelerate testSplitData.c test.c splitLoad.c');
system('./test');

File: examples/Old_tests/test/testSplitData/testSplitData.c
#include <stdio.h>

#include "test.h"

REAL compute_err(REAL *y, REAL *sol, int nn) {
  REAL err = 0.0;
  for(int i=0; i<nn; i++) 
    err += (y[i]-sol[i])*(y[i]-sol[i]);
  return err;
}

int main(void) {
  printf("--- Test matrix-vector multiplication methods ---\n");
  printf("This file tests for correctness of the methods, it does not compare speeds\n");

  // Load the data (which variables are loaded are in test.h)
  loadData();

  REAL y[n];

  printf("\n\n===> Testing matrix-vector products <===\n");

  func_blas(y, x);
  printf("%20s error = %.4e\n", "BLAS", compute_err(y, sol, n));

  func_forloop(y, x);
  printf("%20s error = %.4e\n", "for-loops", compute_err(y, sol, n));

  func_sparse(y, x);
  printf("%20s error = %.4e\n", "Sparse", compute_err(y, sol, n));

  func_exhaustive(y, x);
  printf("%20s error = %.4e\n", "Exhaustive", compute_err(y, sol, n));



  printf("\n\n===> Testing linear solvers <===\n");

  REAL work[x_lin_len];  // Work variable for LDL solve
  REAL y_lin[x_lin_len]; // Solution

  // We copy the input x_lin to a work variable first, since work is modified on exit
  memcpy(work, x_lin, sizeof(REAL)*x_lin_len);
  func_solve_ldl(y_lin, work);
  printf("%20s error = %.4e\n", "LDL", compute_err(y_lin, y_sol, x_lin_len));

  func_solve_inv(y_lin, x_lin);
  printf("%20s error = %.4e\n", "Invert", compute_err(y_lin, y_sol, x_lin_len));

  return 0;
}
